nntrainer/layers/lstm.cpp

```diff
@@ -74,7 +138,72 @@ void LSTMLayer::setRecurrentActivation(ActivationType activation) {
 }
 
 void LSTMLayer::forwarding(bool training) {
-  // NYI
+  Tensor &weight_xh =
+    weightAt(static_cast<int>(LSTMParams::weight_xh)).getVariableRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(LSTMParams::weight_hh)).getVariableRef();
+  Tensor &bias_h =
+    weightAt(static_cast<int>(LSTMParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor &m_cell_ = mem_cell->getVariableRef();
+
+  Tensor temp;
+  Tensor hs_prev;
+  Tensor cs_prev;
+  Tensor hs;
+  Tensor cs;
+  Tensor f;
+  Tensor g;
+  Tensor i;
+  Tensor o;
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor islice = input_.getBatchSlice(b, 1);
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+    Tensor cell = m_cell_.getBatchSlice(b, 1);
+
+    for (unsigned int t = 0; t < islice.height(); ++t) {
+      Tensor xs = input_.getSharedDataTensor(TensorDim(1, 1, 1, islice.width()),
+                                             t * islice.width());
+      hs = oslice.getSharedDataTensor(TensorDim(1, 1, 1, oslice.width()),
+                                      t * oslice.width());
+      cs = cell.getSharedDataTensor(TensorDim(1, 1, 1, cell.width()),
+                                    t * cell.width());
+
+      if (t > 0) {
+        hs_prev = oslice.getSharedDataTensor(TensorDim(1, 1, 1, oslice.width()),
+                                             (t - 1) * oslice.width());
+        cs_prev = cell.getSharedDataTensor(TensorDim(1, 1, 1, cell.width()),
+                                           (t - 1) * cell.width());
+      } else {
+        hs_prev = h_prev.getBatchSlice(b, 1);
+        cs_prev = c_prev.getBatchSlice(b, 1);
+      }
+      hs_prev.dot(weight_hh, temp);
+      temp.add_i(bias_h);
+      temp.add_i(xs.dot(weight_xh));
+
+      f = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 0);
+      g = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), temp.width());
+      i = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 2 * temp.width());
+      o = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 3 * temp.width());
+
+      recurrent_acti_func.run_fn(f, f);
+      recurrent_acti_func.run_fn(i, i);
+      recurrent_acti_func.run_fn(o, o);
+      acti_func.run_fn(g, g);
+
+      f.multiply(cs_prev, cs);
+      cs.add_i(g.multiply_i(i));
+
+      acti_func.run_fn(cs, hs);
+      hs.multiply_i(o);
+    }
+    h_prev.getBatchSlice(b, 1).copy(hs);
```

Comment:
I think it's guaranteed that dimension is same for this case...

If they are different, copy will allocate a new tensor thus breaking memory layout.
I think later, we should have a different memory ready for this purpose.

Reply:
`tensor::copy(tensor)` performs the check to ensure that the size of src and destination tensor are the same.

Reply:
Yes, I mean the case we intend to dump the data without allocation but if we can't it, it would be safer to throw.

Reply:
the dimension is same. I will add checking in later PR.